Error batch replication: runtime error

## NOTE
If this case is urgent, please subscribe to [Subnet](https://min.io/pricing) so that our 24/7 support team may help you faster.

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior

Replication have to start   and replicate my data. 

## Current Behavior
minio client failed  to run replication with this issue: 

```
panic: "POST /minio/admin/v3/start-job": runtime error: invalid memory address or nil pointer dereference
goroutine 124707 [running]:
runtime/debug.Stack()
 runtime/debug/stack.go:24 +0x5e
github.com/minio/minio/cmd.serverMain.func8.setCriticalErrorHandler.func2.1()
 github.com/minio/minio/cmd/generic-handlers.go:555 +0x97
panic({0x289a060?, 0x6425330?})
 runtime/panic.go:914 +0x21f
net/http.(*Transport).roundTrip(0x0, 0xc10dd5c600)
 net/http/transport.go:514 +0x2d
net/http.(*Transport).RoundTrip(0xc10c173340?, 0x4cffe60?)
 net/http/roundtrip.go:17 +0x13
net/http.send(0xc10dd5c600, {0x4cffe60, 0x0}, {0x1?, 0x0?, 0x0?})
 net/http/client.go:260 +0x606
net/http.(*Client).send(0xc10bff4ff0, 0xc10dd5c600, {0x0?, 0x0?, 0x0?})
 net/http/client.go:181 +0x98
net/http.(*Client).do(0xc10bff4ff0, 0xc10dd5c600)
 net/http/client.go:724 +0x912
net/http.(*Client).Do(...)
 net/http/client.go:590


```


## Steps to Reproduce (for bugs)

1. configure replication_batch_file like that: 
```
  replicate:
  apiVersion: v1
  source:
    type: minio 
    bucket: test
    endpoint: "https://minio1-test:20502"
    credentials:
      accessKey: "minioadmin" 
      secretKey: "mypassword" 
    snowball:
       disable:  true
  
  target:
    type: minio
    bucket: test
    endpoint: "https://minio-cluster2-1-test:20512"
    credentials:
      accessKey: "minioadmin" 
      secretKey: "mypassword" 
    
    retry:
      attempts: 10 
      delay: "500ms"  
```
2. Set  alias: `mc alias set local/ https://minio1-test:20502 minioadmin mypassword --insecure`
3.  Launch batch  replication: `mc batch start  local ./barchreplication.yaml`


## Your Environment
* Version used (`minio --version`): 
```
minio  --version
minio version RELEASE.2024-03-10T02-53-48Z (commit-id=88a89213ff6663f2d3ee1de7f3293497ab9844e8)
Runtime: go1.21.8 linux/amd64
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Copyright: 2015-2024 MinIO, Inc.
```

* Server setup and configuration: Red Hat Enterprise Linux release 8.7 (Ootpa)
* Operating System and version (`uname -a`): Linux myVM 4.18.0-425.3.1.el8.x86_64  SMP Fri Sep 30 11:45:06 EDT 2022 x86_64 x86_64 x86_64 GNU/Linux

--- a/cmd/batch-expire.go
+++ b/cmd/batch-expire.go
@@ -321,7 +321,7 @@ func (r BatchJobExpire) Notify(ctx context.Context, body io.Reader) error {
 		req.Header.Set("Authorization", r.NotificationCfg.Token)
 	}
 
-	clnt := http.Client{Transport: getRemoteInstanceTransport}
+	clnt := http.Client{Transport: getRemoteInstanceTransport()}
 	resp, err := clnt.Do(req)
 	if err != nil {
 		return err

--- a/cmd/batch-handlers.go
+++ b/cmd/batch-handlers.go
@@ -92,7 +92,7 @@ func notifyEndpoint(ctx context.Context, ri *batchJobInfo, endpoint, token strin
 	}
 	req.Header.Set("Content-Type", "application/json")
 
-	clnt := http.Client{Transport: getRemoteInstanceTransport}
+	clnt := http.Client{Transport: getRemoteInstanceTransport()}
 	resp, err := clnt.Do(req)
 	if err != nil {
 		return err
@@ -351,7 +351,7 @@ func (r *BatchJobReplicateV1) StartFromSource(ctx context.Context, api ObjectLay
 	c, err := miniogo.New(u.Host, &miniogo.Options{
 		Creds:        credentials.NewStaticV4(cred.AccessKey, cred.SecretKey, cred.SessionToken),
 		Secure:       u.Scheme == "https",
-		Transport:    getRemoteInstanceTransport,
+		Transport:    getRemoteInstanceTransport(),
 		BucketLookup: lookupStyle(r.Source.Path),
 	})
 	if err != nil {
@@ -1048,7 +1048,7 @@ func (r *BatchJobReplicateV1) Start(ctx context.Context, api ObjectLayer, job Ba
 	c, err := miniogo.NewCore(u.Host, &miniogo.Options{
 		Creds:        credentials.NewStaticV4(cred.AccessKey, cred.SecretKey, cred.SessionToken),
 		Secure:       u.Scheme == "https",
-		Transport:    getRemoteInstanceTransport,
+		Transport:    getRemoteInstanceTransport(),
 		BucketLookup: lookupStyle(r.Target.Path),
 	})
 	if err != nil {
@@ -1068,7 +1068,7 @@ func (r *BatchJobReplicateV1) Start(ctx context.Context, api ObjectLayer, job Ba
 			cl, err := miniogo.New(u.Host, &miniogo.Options{
 				Creds:        credentials.NewStaticV4(cred.AccessKey, cred.SecretKey, cred.SessionToken),
 				Secure:       u.Scheme == "https",
-				Transport:    getRemoteInstanceTransport,
+				Transport:    getRemoteInstanceTransport(),
 				BucketLookup: lookupStyle(r.Target.Path),
 			})
 			if err != nil {
@@ -1354,7 +1354,7 @@ func (r *BatchJobReplicateV1) Validate(ctx context.Context, job BatchJobRequest,
 	c, err := miniogo.NewCore(u.Host, &miniogo.Options{
 		Creds:        credentials.NewStaticV4(cred.AccessKey, cred.SecretKey, cred.SessionToken),
 		Secure:       u.Scheme == "https",
-		Transport:    getRemoteInstanceTransport,
+		Transport:    getRemoteInstanceTransport(),
 		BucketLookup: lookupStyle(pathStyle),
 	})
 	if err != nil {

--- a/cmd/config-current.go
+++ b/cmd/config-current.go
@@ -576,11 +576,7 @@ func applyDynamicConfigForSubSys(ctx context.Context, objAPI ObjectLayer, s conf
 
 		globalAPIConfig.init(apiConfig, setDriveCounts)
 		autoGenerateRootCredentials() // Generate the KMS root credentials here since we don't know whether API root access is disabled until now.
-
-		// Initialize remote instance transport once.
-		getRemoteInstanceTransportOnce.Do(func() {
-			getRemoteInstanceTransport = NewHTTPTransportWithTimeout(apiConfig.RemoteTransportDeadline)
-		})
+		setRemoteInstanceTransport(NewHTTPTransportWithTimeout(apiConfig.RemoteTransportDeadline))
 	case config.CompressionSubSys:
 		cmpCfg, err := compress.LookupConfig(s[config.CompressionSubSys][config.Default])
 		if err != nil {

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -34,7 +34,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
-	"sync"
+	"sync/atomic"
 	"time"
 	"unicode"
 
@@ -1291,11 +1291,20 @@ func getCpObjMetadataFromHeader(ctx context.Context, r *http.Request, userMeta m
 	return defaultMeta, nil
 }
 
-// getRemoteInstanceTransport contains a singleton roundtripper.
-var (
-	getRemoteInstanceTransport     *http.Transport
-	getRemoteInstanceTransportOnce sync.Once
-)
+// getRemoteInstanceTransport contains a roundtripper for external (not peers) servers
+var remoteInstanceTransport atomic.Value
+
+func setRemoteInstanceTransport(tr http.RoundTripper) {
+	remoteInstanceTransport.Store(tr)
+}
+
+func getRemoteInstanceTransport() http.RoundTripper {
+	rt, ok := remoteInstanceTransport.Load().(http.RoundTripper)
+	if ok {
+		return rt
+	}
+	return nil
+}
 
 // Returns a minio-go Client configured to access remote host described by destDNSRecord
 // Applicable only in a federated deployment
@@ -1306,7 +1315,7 @@ var getRemoteInstanceClient = func(r *http.Request, host string) (*miniogo.Core,
 	core, err := miniogo.NewCore(host, &miniogo.Options{
 		Creds:     credentials.NewStaticV4(cred.AccessKey, cred.SecretKey, ""),
 		Secure:    globalIsTLS,
-		Transport: getRemoteInstanceTransport,
+		Transport: getRemoteInstanceTransport(),
 	})
 	if err != nil {
 		return nil, err

